jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt

```diff
@@ -0,0 +1,130 @@
+                }
+            }
+        }
+
+        override fun read(dataInput: DataInput): List<IndexedResource> {
+            val size = dataInput.readInt()
+            val resources = mutableListOf<IndexedResource>()
+            repeat(size) {
+                val logicalName = dataInput.readUTF()
+                val type = dataInput.readUTF()
+                val mutableMap: MutableMap<String, String> = mutableMapOf()
+
+                val propertySize = dataInput.readInt()
+                repeat(propertySize) {
+                    val key = dataInput.readUTF()
+                    val value = dataInput.readUTF()
+                    mutableMap[key] = value
+                }
+                resources.add(IndexedResource.from(logicalName, type, mutableMap))
+            }
+            return resources
+        }
+    }
+
+    override fun getName(): ID<String, List<IndexedResource>> = NAME
+
+    override fun getIndexer(): DataIndexer<String, List<IndexedResource>, FileContent> = DataIndexer { fileContent ->
+        val indexedResources = mutableMapOf<String, List<IndexedResource>>()
+
+        fileContent.psiFile.acceptNode(object : PsiElementVisitor() {
+            override fun visitElement(element: PsiElement?) {
+                super.visitElement(element)
+                element?.run {
+                    val parent = element.parent as? YAMLKeyValue ?: return
+                    if (parent.value != this) return
+
+                    val indexedResource = YamlCloudFormationTemplate.convertPsiToResource(parent)?.let {
+                        IndexedResource.fromResource(it)
+                    } ?: return
+
+                    (indexedResources.computeIfAbsent(indexedResource.type) { mutableListOf() } as MutableList)
+                            .add(indexedResource)
+                }
+            }
+        })
+
+        indexedResources
+    }
+
+    override fun getKeyDescriptor(): KeyDescriptor<String> = EnumeratorStringDescriptor.INSTANCE
+
+    override fun getVersion(): Int = 1
+
+    override fun getInputFilter(): FileBasedIndex.InputFilter = fileFilter
+
+    override fun dependsOnFileContent(): Boolean = true
+
+    companion object {
+        private val NAME: ID<String, List<IndexedResource>> = ID.create("SamTemplateIndex")
+
+        private fun PsiElement.acceptNode(visitor: PsiElementVisitor) {
+            accept(visitor)
+
+            if (children.isNotEmpty()) {
+                children.forEach { it.acceptNode(visitor) }
+            }
+        }
+
+        private fun listAllResources(project: Project): Collection<IndexedResource> {
+            val index = FileBasedIndex.getInstance()
+            return index.getAllKeys(NAME, project)
+                    .asSequence()
+                    .mapNotNull { index.getValues(NAME, it, GlobalSearchScope.projectScope(project)) }
+                    .filter { it.isNotEmpty() }
+                    .flatten()
```

Comment:
The `index.getValues` returns a list of `Value`s. In this case, the `Value` type is already a list of resources. That's why I flatten twice since it is kinda 2d array.

Reply:
Ah